File: server/scrape_honda_parts.py
"""
Honda Cengkareng Parts Scraper
Scrapes all spare parts data from Honda Cengkareng website
"""
import requests
from bs4 import BeautifulSoup
import json
import time
import re
from urllib.parse import urljoin

def scrape_page(page_num):
    """Scrape single page of products"""
    url = f"https://www.hondacengkareng.com/kategori-produk/suku-cadang-resmi-motor-honda/page/{page_num}/"
    
    print(f"Scraping page {page_num}...", end=" ")
    
    try:
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.content, 'html.parser')
        products = []
        
        # Find all product links
        product_links = soup.select('a.woocommerce-LoopProduct-link')
        
        for link in product_links:
            product = {}
            
            # Get product name
            h2 = link.select_one('h2.woocommerce-loop-product__title')
            product['name'] = h2.get_text(strip=True) if h2 else 'N/A'
            
            # Get price
            price_elem = link.select_one('.price bdi')
            if price_elem:
                price_text = price_elem.get_text(strip=True)
                # Extract only numbers
                price_clean = re.sub(r'[^\d]', '', price_text)
                product['price'] = int(price_clean) if price_clean else 0
            else:
                product['price'] = 0
            
            # Get SKU/Kode Part
            sku = 'N/A'
            # Look for "Kode Part:" in various p tags
            p_tags = link.select('p')
            for p in p_tags:
                text = p.get_text(strip=True)
                if 'Kode Part:' in text:
                    sku = text.replace('Kode Part:', '').strip()
                    break
            
            product['sku'] = sku
            
            # Get category/group if available
            category = 'HGP'  # Default
            product['group'] = category
            
            # Get unit (default to pcs)
            product['unit'] = 'pcs'
            
            if product['name'] != 'N/A':
                products.append(product)
        
        print(f"✓ Found {len(products)} products")
        return products
        
    except Exception as e:
        print(f"✗ Error: {str(e)}")
        return []

def main():
    print("=" * 60)
    print("Honda Cengkareng Parts Scraper")
    print("=" * 60)
    print()
    
    # Test with first page to see structure
    print("Testing with first page...")
    test_products = scrape_page(1)
    
    if not test_products:
        print("Failed to scrape first page. Exiting.")
        return
    
    print(f"\nSample product from page 1:")
    print(json.dumps(test_products[0], indent=2, ensure_ascii=False))
    print()
    
    # Ask for confirmation
    total_pages = 395  # Based on website
    print(f"Total pages to scrape: {total_pages}")
    print(f"Estimated products: ~{total_pages * 48}")
    print(f"Estimated time: ~{total_pages * 2 / 60:.1f} minutes (assuming 2 sec/page)")
    print()
    
    response = input("Continue with full scrape? (yes/no): ")
    
    if response.lower() != 'yes':
        print("Scraping cancelled.")
        return
    
    print("\nStarting full scrape...")
    print("=" * 60)
    
    all_products = []
    failed_pages = []
    
    for page_num in range(1, total_pages + 1):
        products = scrape_page(page_num)
        
        if products:
            all_products.extend(products)
        else:
            failed_pages.append(page_num)
        
        # Be respectful to server - add delay
        time.sleep(1.5)
        
        # Save progress every 50 pages
        if page_num % 50 == 0:
            with open('honda_parts_progress.json', 'w', encoding='utf-8') as f:
                json.dump(all_products, f, indent=2, ensure_ascii=False)
            print(f"\nProgress saved! Total products so far: {len(all_products)}\n")
    
    # Save final results
    print("\n" + "=" * 60)
    print(f"Scraping completed!")
    print(f"Total products scraped: {len(all_products)}")
    print(f"Failed pages: {len(failed_pages)}")
    
    if failed_pages:
        print(f"Failed page numbers: {failed_pages[:10]}...")
    
    # Save to JSON
    output_file = 'honda_parts_complete.json'
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(all_products, f, indent=2, ensure_ascii=False)
    
    print(f"\nData saved to: {output_file}")
    
    # Generate statistics
    print("\n" + "=" * 60)
    print("Statistics:")
    print(f"- Total products: {len(all_products)}")
    print(f"- Products with SKU: {sum(1 for p in all_products if p['sku'] != 'N/A')}")
    print(f"- Products with price: {sum(1 for p in all_products if p['price'] > 0)}")
    print(f"- Average price: Rp {sum(p['price'] for p in all_products) / len(all_products):,.0f}")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: server/scrape_honda_parts.mjs
/**
 * Honda Cengkareng Parts Scraper - Simple Version
 * No external dependencies, uses built-in Node.js modules only
 */

import https from 'https';
import fs from 'fs';

const BASE_URL = 'https://www.hondacengkareng.com/kategori-produk/suku-cadang-resmi-motor-honda';
const TOTAL_PAGES = 395; // Full scraping - all pages
const DELAY_MS = 2000;

// Simple HTML parser using regex (not perfect but works for this case)
function parseProducts(html) {
    const products = [];

    // Match product blocks - find all product links
    const productPattern = /<a[^>]*class="[^"]*woocommerce-LoopProduct-link[^"]*"[^>]*>([\s\S]*?)<\/a>/g;
    let match;

    while ((match = productPattern.exec(html)) !== null) {
        const productHtml = match[1];

        // Extract name from h2
        const nameMatch = /<h2[^>]*class="[^"]*woocommerce-loop-product__title[^"]*"[^>]*>([^<]+)<\/h2>/.exec(productHtml);
        const name = nameMatch ? nameMatch[1].trim() : 'N/A';

        // Extract price
        const priceMatch = /<bdi[^>]*>([^<]+)<\/bdi>/.exec(productHtml);
        let price = 0;
        if (priceMatch) {
            const priceText = priceMatch[1];
            const priceClean = priceText.replace(/[^\d]/g, '');
            price = priceClean ? parseInt(priceClean) : 0;
        }

        // Extract SKU - look for "Kode Part:"
        let sku = 'N/A';
        const skuMatch = /Kode Part:\s*([A-Z0-9]+)/i.exec(productHtml);
        if (skuMatch) {
            sku = skuMatch[1].trim();
        }

        if (name !== 'N/A') {
            products.push({
                name,
                sku,
                price,
                group: 'HGP',
                unit: 'pcs'
            });
        }
    }

    return products;
}

// Function to fetch HTML content
function fetchPage(pageNum) {
    return new Promise((resolve, reject) => {
        const url = pageNum === 1
            ? `${BASE_URL}/`
            : `${BASE_URL}/page/${pageNum}/`;

        https.get(url, {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        }, (res) => {
            if (res.statusCode !== 200) {
                reject(new Error(`HTTP ${res.statusCode}`));
                return;
            }

            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                resolve(data);
            });

        }).on('error', (err) => {
            reject(err);
        });
    });
}

// Function to sleep/delay
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Main scraping function
async function scrapePage(pageNum) {
    process.stdout.write(`Page ${pageNum}/${TOTAL_PAGES}... `);

    try {
        const html = await fetchPage(pageNum);
        const products = parseProducts(html);
        console.log(`✓ ${products.length} products`);
        return products;
    } catch (error) {
        console.log(`✗ Error: ${error.message}`);
        return [];
    }
}

// Main function
async function main() {
    console.log('='.repeat(60));
    console.log('Honda Cengkareng Parts Scraper');
    console.log('='.repeat(60));
    console.log();

    // Test with first page
    console.log('Testing first page...');
    const testProducts = await scrapePage(1);

    if (testProducts.length === 0) {
        console.log('❌ Failed to scrape first page. Exiting.');
        return;
    }

    console.log('\n📦 Sample product:');
    console.log(JSON.stringify(testProducts[0], null, 2));
    console.log();

    console.log(`📊 Scraping ${TOTAL_PAGES} pages...`);
    console.log(`⏱️  Estimated: ~${(TOTAL_PAGES * DELAY_MS / 1000 / 60).toFixed(1)} minutes`);
    console.log();

    const allProducts = [];
    const failedPages = [];

    for (let pageNum = 1; pageNum <= TOTAL_PAGES; pageNum++) {
        const products = await scrapePage(pageNum);

        if (products.length > 0) {
            allProducts.push(...products);
        } else {
            failedPages.push(pageNum);
        }

        await sleep(DELAY_MS);
    }

    // Save final results
    console.log('\n' + '='.repeat(60));
    console.log('✅ Scraping completed!');
    console.log(`📦 Total products: ${allProducts.length}`);
    console.log(`❌ Failed pages: ${failedPages.length}`);

    const outputFile = 'server/honda_parts_scraped.json';
    fs.writeFileSync(outputFile, JSON.stringify(allProducts, null, 2));

    console.log(`💾 Data saved to: ${outputFile}`);

    // Statistics
    const productsWithSKU = allProducts.filter(p => p.sku !== 'N/A').length;
    const productsWithPrice = allProducts.filter(p => p.price > 0).length;
    const avgPrice = allProducts.reduce((sum, p) => sum + p.price, 0) / allProducts.length;

    console.log('\n📊 Statistics:');
    console.log(`   • Total: ${allProducts.length}`);
    console.log(`   • With SKU: ${productsWithSKU} (${(productsWithSKU / allProducts.length * 100).toFixed(1)}%)`);
    console.log(`   • With price: ${productsWithPrice} (${(productsWithPrice / allProducts.length * 100).toFixed(1)}%)`);
    console.log(`   • Avg price: Rp ${Math.round(avgPrice).toLocaleString('id-ID')}`);
    console.log('='.repeat(60));
    console.log();
    console.log('💡 To scrape ALL 395 pages (18,950 products):');
    console.log('   Edit TOTAL_PAGES to 395 in the script and run again.');
    console.log('   Estimated time: ~13 minutes');
}

// Run
main().catch(err => {
    console.error('Fatal error:', err);
    process.exit(1);
});
